### ISSUE TITLE
offpolicy_trainer tqdm update question

### ISSUE THREAD

**Issue by eric-liuyd** [thu-ml/tianshou#254]
- [x] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [x] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [ ] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, sys
  print(tianshou.__version__, torch.__version__, sys.version, sys.platform)
  ```
Dear author,

I found a bug in tqdm updating in offpolicy_trainer (trainer/offpolicy/offpolicy-trainer)

In line 132 , `t.update(1)` should not be set in for loop. Because this step is for updating policy but not for pushing epoch.

In our experiment, changing 'update_per_step' will shorten the env_step (our episode has fixed env steps), which does not make sense. No matter how 'update_per_step' changes, the total env_step should not change.

So we make a fix here. The original code is as below (line 120 - 133).

```
                  for i in range(update_per_step * min(
                        result["n/st"] // collect_per_step, t.total - t.n)):
                    gradient_step += 1
                    losses = policy.update(batch_size, train_collector.buffer)
                    for k in losses.keys():
                        if stat.get(k) is None:
                            stat[k] = MovAvg()
                        stat[k].add(losses[k])
                        data[k] = f"{stat[k].get():.6f}"
                        if writer and gradient_step % log_interval == 0:
                            writer.add_scalar(
                                k, stat[k].get(), global_step=gradient_step)
                    t.update(1)
                    t.set_postfix(**data)
```

We move  `t.update(1)` out from the for loop and everything is right after. The modified code is as following.

```
                for i in range(update_per_step * int(min(
                        result["n/st"] // collect_per_step, t.total - t.n))):
                    gradient_step += 1
                    losses = policy.update(batch_size, train_collector.buffer)
                    for k in losses.keys():
                        if stat.get(k) is None:
                            stat[k] = MovAvg()
                        stat[k].add(losses[k])
                        data[k] = f"{stat[k].get():.6f}"
                        if writer and gradient_step % log_interval == 0:
                            writer.add_scalar(
                                k, stat[k].get(), global_step=gradient_step)
                    t.set_postfix(**data)
                t.update(1)
```

Please check this issue. Thanks!

**Comment by Trinkle23897** [comment]
I write this `t.update(1)` in order to illustrate each gradient step. That is to say, when a gradient step is done, the tqdm will update once.

> In our experiment, changing 'update_per_step' will shorten the env_step (our episode has fixed env steps), which does not make sense. No matter how 'update_per_step' changes, the total env_step should not change.

I think it is the expected behavior. Let's suppose your episode has a fixed length of 10. If you use `update_per_step=2` and get a total env_step=1000. Once changing to `update_per_step=4`, **the overall gradient step does not change** (1000/2=500), so that at this time the total env step is 2000 (500*4).

**Comment by eric-liuyd** [comment]
> I write this `t.update(1)` in order to illustrate each gradient step. That is to say, when a gradient step is done, the tqdm will update once.
> 
> > In our experiment, changing 'update_per_step' will shorten the env_step (our episode has fixed env steps), which does not make sense. No matter how 'update_per_step' changes, the total env_step should not change.
> 
> I think it is the expected behavior. Let's suppose your episode has a fixed length of 10. If you use `update_per_step=2` and get a total env_step=1000. Once changing to `update_per_step=4`, **the overall gradient step does not change** (1000/2=500), so that at this time the total env step is 2000 (500*4).

Sorry, I confused the 'step_per_epoch' in your function with "the numbers of env.step() running", which leads to incorrect results. I'll close it.

**Comment by Zhikaiiii** [comment]
Hey, I have a question. `t.total` is equal to `step_per_epoch`, but `t.update(1)` may be called much more time because of `update_per_step` so the output of tqdm may be a little strange when exceed `step_per_epoch`.
Also, I wonder if there is some way to ignore the output of tqdm because I only want test result recorded in my log, thanks!

**Comment by Trinkle23897** [comment]
> Hey, I have a question. `t.total` is equal to `step_per_epoch`, but `t.update(1)` may be called much more time because of `update_per_step` so the output of tqdm may be a little strange when exceed `step_per_epoch`.

Good question. You are right because currently we cannot stop the network update in exactly `step_per_epoch`, especially for on-policy algorithm (this issue for off-policy algorithm will be fixed these days, see #274 for details).

> Also, I wonder if there is some way to ignore the output of tqdm because I only want test result recorded in my log, thanks!

This can be done by output redirect, i.e., `python3 xxx.py > log` (because tqdm output is in stderr).
